[18.06,19.07] build: fixup python SetupHostCommand to use python2

champtar opened the pull request:
~~On some distributions (Fedora 31 at least), python now points to python3
I did a full build of ath79 in a Fedora 31 container without `python` in the PATH (only `python2` / `python3`) and the only python related failure that I see is mailman that explicitly uses `/usr/bin/python`~~

This replaces #2530

Edit: 
Here is a way to break your build env without this patch:
1) have python point to python3, and no python2
2) start the build, SetupHostCommand will create a symlink
./staging_dir/host/bin/python -> /usr/bin/python
3) build fails on scons because it can't find any python2
4) install python2 and restart the build
5) the build fails on wireless-regdb compile because python is python3 instead of python2

champtar committed:
build: fixup python SetupHostCommand to use python2

Here is a way to break your build env without this patch:
1) have python point to python3, and no python2
2) start the build, SetupHostCommand will create a symlink
./staging_dir/host/bin/python -> /usr/bin/python
3) build fails on scons because it can't find any python2
4) install python2 and restart the build
5) the build fails on wireless-regdb compile because python is python3 instead of python

Signed-off-by: Etienne Champetier <champetier.etienne@gmail.com>

ynezz commented on include/prereq-build.mk:
```diff
@@ -142,9 +142,15 @@ $(eval $(call SetupHostCommand,perl,Please install Perl 5.x, \
+	python2 -V 2>&1 | grep 'Python 2', \
+	python -V 2>&1 | grep 'Python 2'))
+
+$(eval $(call SetupHostCommand,python3,Please install Python >= 3.5, \
```
This should go into separate commit, it's not clear why do you need this for 19.07 which is python2 only.

champtar replied:
I want to put python3 as a requirement so that people can use it in their packages updates, but yes separate commit with an explanation is needed

ynezz replied:
> I want to put python3 as a requirement

Don't waste time on that, it has zero chances to be accepted.

It has all started, because you had no python2 by default installed in your Fedora31, now you're going to add python3 requirement for everybody even if it's not needed by anything in the tree? If it's needed by some package in the feed, that please handle it there, at the proper place.

ynezz commented on rules.mk:
```diff
@@ -276,7 +276,9 @@ BASH:=bash
-PYTHON:=python
+PYTHON:=python2
+PYTHON2:=python2
+PYTHON3:=python3
```
There is no use for this variable.

ynezz commented on package/firmware/layerscape/ls-rcw/patches/0001-rcw-support-byte-swapping-without-tclsh-tool.patch:
```diff
@@ -32,7 +32,7 @@ index 0000000..386310e
 +++ b/byte_swap.py
 @@ -0,0 +1,32 @@
-+#!/usr/bin/env python
++#!/usr/bin/env python2
```
It doesn't make sense to me, as SetupHostCommand is going to create proper python symlink in the staging/host/bin directory which is then first in the path, so it should work as it is.

champtar replied:
I need to have a second look, maybe SetupHostCommand is broken in some cases then. I started because build broke for me on Fedora 31

champtar replied:
The issue is that if `./staging_dir/host/bin/python` already exists, SetupHostCommand doesn't seem to update it when it rerun

ynezz commented on include/prereq-build.mk:
```diff
@@ -142,9 +142,15 @@ $(eval $(call SetupHostCommand,perl,Please install Perl 5.x, \
-	python2.7 -V 2>&1 | grep Python, \
-	python2 -V 2>&1 | grep Python, \
-	python -V 2>&1 | grep Python))
+	python2.7 -V 2>&1 | grep 'Python 2.7', \
```
This makes sense, should be in separate commit, with `build:` subject prefix.

champtar replied:
Done

ynezz commented on package/firmware/wireless-regdb/Makefile:
```diff
@@ -20,7 +20,7 @@ define Package/wireless-regdb
 
 define Build/Compile
-	python $(PKG_BUILD_DIR)/db2fw.py $(PKG_BUILD_DIR)/regulatory.db $(PKG_BUILD_DIR)/db.txt
+	$(PYTHON2) $(PKG_BUILD_DIR)/db2fw.py $(PKG_BUILD_DIR)/regulatory.db $(PKG_BUILD_DIR)/db.txt
```
Backporting b6bae4a2c9f11f7e55319c2b4c709396ce649688  would probably make sense here.

ynezz replied:
Either way, this should be just $(PYTHON) not PYTHON2 explicitly, as python2 is implicit in 19.07.

champtar replied:
https://github.com/openwrt/openwrt/commit/b6bae4a2c9f11f7e55319c2b4c709396ce649688 is an ugly way to fix python2 python3 confusion, we should just stop using python without the version number, it makes everything easier to understand and everything working

ynezz commented on rules.mk:
```diff
@@ -276,7 +276,9 @@ BASH:=bash
 FIND:=find
 PATCH:=patch
-PYTHON:=python
+PYTHON:=python2
```
This seems wrong as SetupHostCommand is going to create `python` symlink, not `python2` symlink.

champtar replied:
I should add a python2 SetupHostCommand then

ynezz commented on rules.mk:
```diff
@@ -276,7 +276,9 @@ BASH:=bash
 PATCH:=patch
-PYTHON:=python
+PYTHON:=python2
+PYTHON2:=python2
```
This variable is not needed, python2 is implicit.

champtar replied:
My idea was to make it explicit and to stop using PYTHON var

champtar commented:
Thanks @ynezz for the review, I will re-reproduce my build failure in a clean environment, indeed SetupHostCommand should make everything work

ynezz commented:
> re-reproduce my build failure in a clean environment

Indeed, it would really help if you could provide some minimal Docker based test commands/script in order to understand the root of the problem properly.

champtar commented:
I found the issue, at some point in my fedora 31 update I removed python2, did a clean OpenWRT master, it worked, then started a clean 19.07 build, it failed, so I reinstalled python2 but the SetupHostCommand symlink already pointed to /usr/bin/python (python3), then I started to patch everything instead of looking at why "SetupHostCommand was not working"

If wanted I can add a commit with but it should not be needed except for people switching between master and 19.07
```
define CleanupPython3
  define Require/python3-cleanup
    if [ -f "$(STAGING_DIR_HOST)/bin/python" ] && \
        $(STAGING_DIR_HOST)/bin/python -V 2>&1 | \
        grep -q 'Python 3'; then \
            rm $(STAGING_DIR_HOST)/bin/python; \
    fi
  endef

  $$(eval $$(call Require,python3-cleanup))
endef
```

champtar commented:
This can also be cherry-picked to 18.06

ynezz commented:
> If wanted I can add a commit with but it should not be needed except for people switching between master and 19.07

It makes sense, yes.

champtar commented:
@ynezz for https://github.com/openwrt/openwrt/pull/2533/commits/4e135c7b3c8d2c629cbf1c07c061d8d9e13bbbac as it's a cherry-pick of https://github.com/openwrt/openwrt/commit/3b68fb57c938af3948ae4c2da61501183fbef649 just replacing 2 with 3 I left you as git author and signed-off-by, tell me if it's ok

champtar committed:
build: cleanup possibly dangling Python 3 host symlink

When switching from master branch to 19.07 or older, we need to ensure
that Python symlink in staging bin directory points to Python 2.

We can't rely completly just on SetupHostCommand as its executed only in
cases when the $(STAGING_DIR_HOST)/bin/python doesn't already exist, so
we need to remove it before running SetupHostCommand.

This is a cherry-pick of 3b68fb57c938af3948ae4c2da61501183fbef649
with python3 instead of python2

Signed-off-by: Petr Štetiar <ynezz@true.cz>
Signed-off-by: Etienne Champetier <champetier.etienne@gmail.com>

ynezz commented:
Merged, thanks!
